src/lib/rng/stateful_rng/stateful_rng.cpp

```diff
@@ -34,68 +34,47 @@ bool Stateful_RNG::is_seeded() const
    return m_reseed_counter > 0;
    }
 
-void Stateful_RNG::add_entropy(const uint8_t input[], size_t input_len)
-   {
-   lock_guard_type<recursive_mutex_type> lock(m_mutex);
-
-   update(input, input_len);
-
-   if(8*input_len >= security_level())
-      {
-      reset_reseed_counter();
-      }
-   }
-
-void Stateful_RNG::initialize_with(const uint8_t input[], size_t len)
+void Stateful_RNG::initialize_with(std::span<const uint8_t> input)
    {
    lock_guard_type<recursive_mutex_type> lock(m_mutex);
 
    clear();
-   add_entropy(input, len);
-   }
-
-void Stateful_RNG::randomize(uint8_t output[], size_t output_len)
-   {
-   randomize_with_input(output, output_len, nullptr, 0);
+   add_entropy(input);
    }
 
-void Stateful_RNG::randomize_with_ts_input(uint8_t output[], size_t output_len)
+void Stateful_RNG::fill_bytes_with_input(std::span<uint8_t> output, std::span<const uint8_t> input)
    {
-   uint8_t additional_input[20] = { 0 };
-
-   store_le(OS::get_high_resolution_clock(), additional_input);
+   lock_guard_type<recursive_mutex_type> lock(m_mutex);
 
-#if defined(BOTAN_HAS_SYSTEM_RNG)
-   System_RNG system_rng;
-   system_rng.randomize(additional_input + 8, sizeof(additional_input) - 8);
-#else
-   store_le(OS::get_system_timestamp_ns(), additional_input + 8);
-   store_le(OS::get_process_id(), additional_input + 16);
-#endif
+   if(output.empty())
+      {
+      // We won't produce any output but we must give the derived stateful
+      // RNG a chance to use the provided input for seeding. Therefore, no
+      // reseed_check() is performed before calling generate_output().
+      this->generate_output({}, input);
```

Comment:
This changes the behavior from before:

1. A call to `randomize_with_input()` then no output was requested was a noop before. I think it is good to change this, as the default implementation did behave different. But still something to keep in mind.
2. When calling `randomize_with_input()` with both input and output, we now do this additional `generate_output({}, input)`, even if the input will be later already included via the other calls to `generate_output()`. Is this intentionally? The biggest problem I see here is that e.g. HMAC would no longer follow strictly *NIST SP800-90A section 10.1.2.2*. Even if the additional update should not hurt, not following the specification without a good reason is I think something we should avoid. 
3. Calling `add_entropy()` for a RNG that has no limit (i.e. `max_number_of_bytes_per_request() == 0`, which at least for HMAC should never be the case) will result in a call to `reseed_check()` and `generate_output()`. Is this something we want?

Reply:
> A call to randomize_with_input() then no output was requested was a noop before. I think it is good to change this, as the default implementation did behave different. But still something to keep in mind.

I agree with that. Input would now be incorporated. I also agree that this is probably fine and should likely have been the case before.

Re 2:

If `output` is non-empty we won't enter this branch (line 49: `if(output.empty()) // ...`). Otherwise this would indeed be a problem. There are KAT-tests for HMAC DRBG that back this implementation and I would expect them to fail if we'd deviate from the spec like that.

Re 3:

We wouldn't want that, and it won't happen either. I believe you missed the `return` statement (line 61) that early-returns from this special-case branch. I.e. for `output.empty()` this special case will only update the underlying implementation's state but won't actually try to draw any random bits.

Nevertheless, I think this discussion is making the point that the control flow in this method is actually too hard to disect. That's what I meant with "increases the control flow complexity" [here](https://github.com/randombit/botan/pull/3195/files#r1121292472).

Reply:
2) You are right. I think for some reason I thought of the `if(output.empty())` being `if(!input.empty())` then I wrote this.

3) Yes I misses the `return` completely.

And agree that it is not the easiest to follow how the `Stateful_RNG::fill_bytes_with_input()` and `generate_output()` interact now.
Don't know if keeping the `Stateful_RNG::add_entropy()` and `Stateful_RNG::randomize_with_input()` as private methods would help here or if that is just more indirection without real benefit. I.e.:

```c++
void Stateful_RNG::fill_bytes_with_input(std::span<uint8_t> output, std::span<const uint8_t> input)
   {
   if(output.empty())
      {
      this->add_entropy(input);
      return;
      }
   this->randomize_with_input(output, input);
   }
```

But I think a lot of the problem came also from me trying to understand it just viewing it in the unified diff view.

Reply:
I'll add a bit more documentation to the control flow in the hope that this will be enough.